tensorboard/components/vz_bar_chart/vz-bar-chart.ts

```diff
@@ -0,0 +1,310 @@
+    // We directly reference properties of `this` because this call is
+    // asynchronous, and values may have changed in between the call being
+    // initiated and actually being run.
+    var chart =
+        new BarChart(this.data, this.colorScale, tooltip, this.tooltipColumns);
+    var div = d3.select(this.$.chartdiv);
+    chart.renderTo(div);
+    this._chart = chart;
+  },
+  _computeTooltipTableHeaderHtml: function(tooltipColumns) {
+    // The first column contains the circle with the color of the run.
+    var titles = [''].concat(_.map(tooltipColumns, 'title'));
+    return titles
+        .map(function(title) {
+          return '<th>' + title + '</th>';
+        })
+        .join('');
+  },
+});
+
+class BarChart {
+  private data: {[key: string]: (Bar[])};
+  private colorScale: Plottable.Scales.Color;
+  private tooltip: d3.Selection<any, any, any, any>;
+  private outer: Plottable.Components.Table;
+  private plot: Plottable.Plots.ClusteredBar<string, number>;
+  private targetSVG: d3.Selection<any, any, any, any>;
+
+  constructor(
+      data: {[key: string]: (Bar[])},
+      colorScale: Plottable.Scales.Color,
+      tooltip: d3.Selection<any, any, any, any>,
+      tooltipColumns: vz_chart_helpers.TooltipColumn[]) {
+    // Assign each class a color.
+    colorScale.domain(_.sortBy(_.keys(data)));
+
+    // Assign arguments passed in constructor for future use.
+    this.data = data;
+    this.colorScale = colorScale;
+    this.tooltip = tooltip;
+
+    this.plot = null;
+    this.outer = null;
+
+    // Do things to actually build the chart.
+    this.buildChart(data, colorScale);
+    this.setupTooltips(tooltipColumns);
+  }
+
+  private buildChart(
+      data: {[key: string]: (Bar[])},
+      colorScale: Plottable.Scales.Color) {
+    if (this.outer) {
+      this.outer.destroy();
+    }
+
+    const xScale = new Plottable.Scales.Category();
+    const yScale = new Plottable.Scales.Linear();
+
+    const xAxis = new Plottable.Axes.Category(xScale, 'bottom');
+    const yAxis = new Plottable.Axes.Numeric(yScale, 'left');
+
+    const plot = new Plottable.Plots.ClusteredBar<string, number>();
+    plot.x(function(d) {
+      return d.x;
+    }, xScale);
+    plot.y(function(d) {
+      return d.y;
+    }, yScale);
+
+    const seriesNames = _.keys(data);
+    seriesNames.forEach(
+        seriesName => plot.addDataset(
+            new Plottable.Dataset(data[seriesName]).metadata(seriesName)));
+    plot.attr('fill', function(d, i, dataset) {
+      return colorScale.scale(dataset.metadata());
+    });
+
+    this.plot = plot;
+    this.outer = new Plottable.Components.Table(
+        [[yAxis, plot], [null, xAxis]]);
+  }
+
+  private setupTooltips(tooltipColumns: vz_chart_helpers.TooltipColumn[]) {
+    const plot = this.plot;
+    const pointer = new Plottable.Interactions.Pointer();
+    pointer.attachTo(plot);
+
+    var hideTooltips = () => {
+      this.tooltip.style('opacity', 0);
```

Comment:
Just curious, is there a particular reason we use opacity 0 plus pointer-events none rather than e.g. display none or visibility hidden?  If we ever added interactivity (like #1019 is asking for) we'd need to change it.

Reply:
Well `opacity: 0` by default leaves elements interactive (e.g. links can still be clicked even though you can't see them).  We also apparently set `pointer-events: none` on the tooltip div to work around this, which makes it unclickable, as well as `cursor: none` so there's no cursor changes.

If we wanted to make it possible to "pin" the tooltip or even interact with it (like selecting text or clicking download links) we would have to undo `pointer-events: none` and `cursor: none`, at least while the tooltip is being displayed.  Overall that seems quite a bit more complicated than just toggling `display: none` so I was curious if there was some reason we used opacity instead.

Reply:
Ah, indeed. Yeah, I wonder what motivates `pointer-events: none` here. Pity the original dev didn't leave comments. It seems irrelevant since users can't hover over the tooltip anyway.

My best guess for why `opacity: 0` + `pointer-events: none` is that someone wanted to bypass some Plottable API that detected hidden elements via display or visibility.

Reply:
Hmm, that seems like a reasonable guess.  FWIW though `pointer-events` is not irrelevant - when the tooltip has opacity 0 it still occupies the same space, it's literally like it's there but just transparent.  Without `pointer-events: none` you wouldn't be able to click through the space that the tooltip occupies when it's rendered, even when it was invisible, because you'd be clicking the tooltip div instead of whatever was under it (typically another chart).

Reply:
Ah, true!